README.md
xmonad
xmobar
stalonetray

###################################


history
fromList [("Run or Raise: ",["mconfig","","mate-panel","y","exit"]),("Run: ",["nm-applet --sm-disable","","gnome-screenshot-command --lock","timedatectl","timedatectl\r\b\b\b\b\b"]),("XMonad: ",["restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm-no-resume","restart-wm","restart-wm","restart-wm","restart-wm-no-resume","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm-no-resume","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","xloadimage","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","restart-wm","next-layout","xterm","view\"2\"","view\"1\"","restart-wm","restart-wm"])]

org.xmonad.hs
import XMonad
--------------------------
import System.IO
import System.Exit
--------------------------
import XMonad.Hooks.FadeInactive
import XMonad.Hooks.ManageDocks                                                                  
import XMonad.Hooks.ManageHelpers
import XMonad.Hooks.SetWMName
import XMonad.Hooks.DynamicLog
-----------------------------------------
import XMonad.Layout.ResizableTile
import XMonad.Layout.NoBorders
import XMonad.Layout.SimplestFloat
import XMonad.Layout.Tabbed
import XMonad.Layout.IM
import XMonad.Layout.Fullscreen
import XMonad.Layout.Spiral
import XMonad.Layout.Grid  
-----------------------------------------
import Graphics.X11.Xlib
import XMonad.Prompt
import XMonad.Prompt.Shell
import XMonad.Prompt.XMonad
import XMonad.Util.EZConfig(additionalKeys)
import Graphics.X11.ExtraTypes.XF86
import XMonad.Util.Run(spawnPipe)
--------------------------------------------
import Data.List
import Data.Map (fromList)
import XMonad.Actions.WindowGo
--------------------------------------------
import qualified XMonad.StackSet as W                                                                
import qualified Data.Map as M

-- Manual adding modules @ ~/.xmonad/lib/XMonad
------------------------------------
myModMask :: KeyMask
myTerminal      = "gnome-terminal"
--myTerminal	= "xterm"
myModMask       = mod4Mask --Win-icon key
myBorderWidth   = 2 
myManagementHooks :: [ManageHook]
myLogHook :: X ()
-------------------------------------------------------------------------------------------------------------------------------------
myLogHook = fadeInactiveLogHook fadeAmount
  where
        fadeAmount = 0.5
myWorkspaces = ["1:main","2:web","3:code","4:dev","5:multi"] ++ map show [6..7]
-- - See more at: http://www.tonicebrian.com/2011/09/05/my-working-environment-with-xmonad/#sthash.naKWmCyr.dpuf
--------------------------
myManagementHooks =  
	[
               	resource =? "stalonetray" --> doIgnore,
		className=? "Google-chrome-stable" --> doShift "2:web",
		stringProperty "WM_WINDOW_ROLE" =? "pop-up" --> doCenterFloat,
		(className =? "Firefox" <&&> title =? "ManageEngine ServiceDesk Plus - View Requester Details - Mozilla Firefox") -->  doCenterFloat,
	--	stringProperty "WM_WINDOW_ROLE" =? "browser" --> doCenterFloat,
		className =? "Firefox" --> doShift "2:web",
		className=? "nxclient" --> doShift "4:dev",
               	className=? "Remmina" --> doShift "4:dev",
               	className=? "P4v.bin" --> doShift "3:code",
               	className=? "banshee" --> doShift "5:multi",
               	className=? "VLC" --> doShift "5:multi",
	       	className=? "Pidgin" --> doShift "1:main",
		resource =? "xcalc" --> doCenterFloat,
		className =? "Xmessage" --> doCenterFloat, 
		resource =? "Dialog" --> doCenterFloat,
		className =? "File Operation Progress"   --> doFloat
	]
--Main config
main = do
    spawn "ibus-daemon"
    spawn "xautolock"
    spawn "gnome-settings-daemon"
    spawn "xcompmgr -c"
    spawn "nm-applet --sm-disable"
    spawn "stalonetray"
    spawn "feh --bg-scale ~/.wallpapers/desktop.jpg"
    spawn "/opt/dropbox/dropbox &"
    xmproc <- spawnPipe "/usr/bin/xmobar ~/.xmonad/xmobarrc"
    xmonad $ defaultConfig { 
	manageHook = manageDocks <+> manageHook defaultConfig <+> composeAll myManagementHooks 
        , layoutHook = avoidStruts  $ smartBorders $ myLayout 
        , logHook = myLogHook <+> dynamicLogWithPP xmobarPP
                { ppOutput = hPutStrLn xmproc
                , ppTitle = xmobarColor "blue" "" . shorten 50
                }
        , terminal    	= myTerminal
        , modMask     	= myModMask
        , borderWidth 	= myBorderWidth
        , workspaces 	= myWorkspaces
        , keys 		= myKeys <+> keys defaultConfig
        , focusFollowsMouse = False
  }
	
--myLayout = tiled ||| Mirror tiled ||| Full 
--      where 
--       	-- tiled = Tall 1 (3/100) (3/5)
--	-- big master
--	   tiled = Tall 1 (3/100) (7/11)
myLayout = avoidStruts (
    	Tall 1 (3/100) (1/2) |||
   	Mirror (Tall 1 (3/100) (1/2)) |||
    	tabbed shrinkText tabConfig |||
	--tabbed |||
    	Full |||
    	spiral (6/7)) -- |||
	--Full
    --noBorders (fullscreenFull Full)
------------------------------------------------------------
tabConfig = defaultTheme 
	{ activeBorderColor = "#7C7C7C"
	, activeTextColor = "#EE9A00"
	, activeColor = "#222222"
	, inactiveBorderColor = "#7C7C7C"
	, inactiveTextColor = "#bbbbbb"
	, inactiveColor = "#222222"
		}
-- Binding keys setting
myKeys conf@(XConfig {XMonad.modMask = myModMask}) = M.fromList
        [ ((myModMask, xK_F12), 	xmonadPrompt 	defaultXPConfig)
        , ((myModMask, xK_F3 ), 	shellPrompt  	defaultXPConfig)
	, ((myModMask,            	xK_n), 		spawn "nautilus")
	, ((myModMask, 			xK_c ),		spawn "calibre")
	, ((myModMask .|. controlMask,  xK_c),          spawn "xchat")
	, ((myModMask .|. controlMask,	xK_e),          spawn "evince")
	, ((myModMask,  	        xK_x), 		spawn "nxclient")
	, ((myModMask,			xK_o),		spawn "google-chrome-stable")
	, ((myModMask,			xK_s),		spawn "stardict")
	, ((myModMask,			xK_i), 		spawn "pidgin")
        , ((myModMask,                  xK_r),          spawn "remmina")
        , ((myModMask,                  xK_f),          spawn "p4v")
	, ((myModMask,                  xK_comma), 	sendMessage (IncMasterN 1))
	, ((myModMask,                  xK_period), 	sendMessage (IncMasterN (-1)))
        , ((myModMask,                  xK_b),          spawn "ebook-viewer")
	, ((myModMask,                  xK_Print),      spawn "/home/tduong/.xmonad/bin/select-sreenshot")
	, ((0,                          xK_Print),      spawn "/home/tduong/.xmonad/bin/screenshot") 
        --, ((myModMask,                  xK_Print),      spawn "shutter -s")
	--, ((myModMask,	 		xK_Print), 	spawn "sleep 0.2; scrot -s")
	--, ((0, 				xK_Print), 	spawn "scrot")	
      --  , ((0, 				xK_Print), 	spawn "shutter -f")
        , ((myModMask,                  xK_g),          spawn "/opt/smartgit/bin/smartgit")
        , ((myModMask,                  xK_q),          spawn "killall stalonetray xmobar nm-applet gnome-setting-daemon xautolock" >> restart "xmonad" True)
 	--Volume control
	, ((0 , xF86XK_AudioLowerVolume),               spawn "amixer -c0 set Master on && amixer -c0 set Headphone on && amixer -c0 set Master 5%-")
	, ((0 , xF86XK_AudioRaiseVolume),               spawn "amixer -c0 set Master on && amixer -c0 set Headphone on && amixer -c0 set Master 5%+")
	, ((0 , xF86XK_AudioMute), 			spawn "amixer set Master toggle && amixer set Headphone toggle")
	-----------------------------
        , ((myModMask .|. controlMask,   xK_l),          spawn "xscreensaver-command -lock")
        , ((myModMask,                   xK_v),          spawn "/opt/cisco/anyconnect/bin/vpnui &")
        , ((myModMask .|.controlMask,    xK_v),         spawn "wine '/home/tduong/.wine/drive_c/Program Files (x86)/Visual CertExam Suite/manager.exe'")
        ]
--- Added more application "ipcalc, eog, scribus, inkscape, evince,


lib/XMonad/Layout/Transparent.hs
{-# LANGUAGE MultiParamTypeClasses, TypeSynonymInstances, FlexibleContexts, DeriveDataTypeable, PatternGuards #-}
----------------------------------------------------------------------------
-- |
-- Module      :  XMonad.Layout.Transparent
--
-- Stability   :  unstable
-- Portability :  not portable
--
-- Makes the windows transparent
--
-----------------------------------------------------------------------------

module XMonad.Layout.Transparent (
        withTransp,
        TransparentMsg(..),
    ) where

import XMonad
import qualified XMonad.StackSet as W
import XMonad.Layout.LayoutModifier
import XMonad.Hooks.FadeInactive (fadeOut)
import Data.List

data TransparentMsg = SetOpacity Rational deriving (Show, Read, Typeable)
instance Message TransparentMsg

data Transparent a = Transparent Rational deriving (Show, Read)
instance LayoutModifier Transparent Window where
    modifyLayout (Transparent opa) = applyTransp opa

    pureMess (Transparent opa) m
      | Just (SetOpacity o) <- fromMessage m =
         Just $ Transparent o
      | otherwise = Nothing

applyTransp :: (LayoutClass l Window) =>
               Rational
            -> W.Workspace WorkspaceId (l Window) Window
            -> Rectangle
            -> X ([(Window, Rectangle)], Maybe (l Window))
applyTransp opa wksp rect = do
    let stack = W.stack wksp
    let ws = W.integrate' $ stack
    mapM_ (fadeOut opa) ws
    runLayout wksp rect

withTransp :: Rational -> l a -> ModifiedLayout Transparent l a
withTransp opa = ModifiedLayout $ Transparent opa


xmonad.hs
import XMonad
--------------------------
import System.IO
import System.Exit
--------------------------
import XMonad.Hooks.FadeInactive
import XMonad.Hooks.ManageDocks                                                                  
import XMonad.Hooks.ManageHelpers
import XMonad.Hooks.SetWMName
import XMonad.Hooks.DynamicLog
-----------------------------------------
import XMonad.Layout.ResizableTile
import XMonad.Layout.NoBorders
import XMonad.Layout.SimplestFloat
import XMonad.Layout.Tabbed
import XMonad.Layout.IM
import XMonad.Layout.Fullscreen
import XMonad.Layout.Spiral
import XMonad.Layout.Grid  
-----------------------------------------
import Graphics.X11.Xlib
import XMonad.Prompt
import XMonad.Prompt.Shell
import XMonad.Prompt.XMonad
import XMonad.Util.EZConfig(additionalKeys)
import Graphics.X11.ExtraTypes.XF86
import XMonad.Util.Run(spawnPipe)
--------------------------------------------
import Data.List	(isPrefixOf, isInfixOf)
import Data.Map (fromList)
import XMonad.Actions.WindowGo
--------------------------------------------
import qualified XMonad.StackSet as W                                                                
import qualified Data.Map as M
------------------------------------
import XMonad.Hooks.EwmhDesktops

myModMask :: KeyMask
myTerminal      = "gnome-terminal"
--myTerminal	= "xterm"
myModMask       = mod4Mask --Win-icon key
myBorderWidth   = 2 
myManagementHooks :: [ManageHook]
myLogHook :: X ()
--handleEventHook = ewmhDesktopsEventHook
--logHook = ewmhDesktopsLogHook
-------------------------------------------------------------------------------------------------------------------------------------
myLogHook = fadeInactiveLogHook fadeAmount
  where
        --fadeAmount = 0.5
	fadeAmount = 0.7
myWorkspaces = ["1:main","2:web","3:code","4:remote","5:multi","6:train","7:ref", "8:media", "9:scripting"] -- ++ map show [7..9]
-- See more at: http://www.tonicebrian.com/2011/09/05/my-working-environment-with-xmonad/#sthash.naKWmCyr.dpuf
--------------------------
myManagementHooks =  
	[
               	resource =? "stalonetray" 			--> doIgnore,
		fmap ( "libreoffice" `isPrefixOf` ) className	--> doShift "5:multi",
		className=? "Google-chrome-stable" 		--> doShift "2:web",
		stringProperty "WM_WINDOW_ROLE" =? "pop-up" 	--> doCenterFloat,
		className =? "Firefox" 	--> doShift "2:web",
		fmap ("ManageEngine ServiceDesk Plus - Send Notification" `isPrefixOf`) title --> doCenterFloat,
		fmap ("ManageEngine ServiceDesk Plus - View Requester Details" `isPrefixOf`) title --> doCenterFloat,
		className=? "nxclient" 	--> doShift "4:remote",
		fmap ("NX" `isPrefixOf`) title --> doShift "4:remote",
		stringProperty "WM_WINDOW_ROLE" =? "LoginWidget"	--> doCenterFloat,
               	className=? "Remmina" 	--> doShift "4:remote",
		className =? "Vncviewer" --> doShift "4:remote",
		appName=? "VNC Viewer" --> doShift "4:remote",
               	className=? "P4v.bin" 	--> doShift "3:code",
               	className=? "banshee" 	--> doShift "5:multi",
		appName=? "manager.exe" --> doShift "6:train",
		appName=? "vdpau" 	--> doShift "8:media",
               	className=? "VLC" 	--> doShift "8:media",
		className =? "MPlayer"	--> doShift "8:media",
	       	stringProperty "WM_WINDOW_ROLE" =? "conversation" --> doCenterFloat,
		className=? "Pidgin"	--> doShift "1:main",
		--stringProperty "WM_WINDOW_ROLE" =? "conversation" --> doCenterFloat,
		resource =? "xcalc" 	--> doCenterFloat,
		className =? "Xmessage" --> doCenterFloat, 
		resource =? "Dialog" 	--> doCenterFloat,
		className =? "xpad" 	--> doCenterFloat,
	--	fmap ( "Skype" `isInfixOf` ) className          --> doCenterFloat <+> doShift "1:main",
		className=? "Skype"	--> doShift "1:main",
		stringProperty "WM_WINDOW_ROLE" =? "ConversationsWindow"	--> doCenterFloat <+> doShift "1:main",
		className =? "File Operation Progress"   	--> doFloat,
		fmap ("NetApp" `isPrefixOf`) title --> doShift "4:remote" <+> doCenterFloat
	]
myLayout = avoidStruts (
    	Tall 1 (3/100) (1/2) |||
   	Mirror (Tall 1 (3/100) (1/2)) |||
    	tabbed shrinkText tabConfig |||
	--tabbed |||
    	Full |||
    	spiral (6/7)) -- |||
	--Full
    --noBorders (fullscreenFull Full)
------------------------------------------------------------
tabConfig = defaultTheme { 
	  activeBorderColor = "#7C7C7C"
	, activeTextColor = "#EE9A00"
	, activeColor = "#222222"
	, inactiveBorderColor = "#7C7C7C"
	, inactiveTextColor = "#bbbbbb"
	, inactiveColor = "#222222"
		}
-- Binding keys setting
myKeys conf@(XConfig {XMonad.modMask = myModMask}) = M.fromList
        [ ((myModMask, xK_F12), 	xmonadPrompt 	defaultXPConfig)
        , ((myModMask, xK_F3 ), 	shellPrompt  	defaultXPConfig)
	, ((myModMask,                  xK_a),          spawn "google-chrome-stable")
	, ((myModMask .|. shiftMask,    xK_a),          spawn "chromium")
	, ((myModMask,                  xK_b),          spawn "ebook-viewer")
	, ((myModMask, 			xK_c),		spawn "calibre")
	, ((myModMask .|. controlMask,  xK_c),          spawn "xchat")
	, ((myModMask .|. mod1Mask,	xK_c), 		spawn "xcalc")
	, ((myModMask,                  xK_d),          spawn "gimp") 
	, ((myModMask .|. controlMask,	xK_e),          spawn "evince")
	, ((myModMask,                  xK_f),          spawn "firefox")
        , ((myModMask .|. controlMask,  xK_f),          spawn "p4v")
	, ((myModMask,                  xK_g),          spawn "/usr/bin/smartgithg >& /dev/null")
	, ((myModMask,                  xK_i),          spawn "pidgin") 
	, ((myModMask,			xK_o),		spawn "libreoffice")
	, ((myModMask,			xK_s),		spawn "stardict")
	, ((myModMask .|. shiftMask,	xK_s),		spawn "skype")
	, ((myModMask,                  xK_n),          spawn "java -jar /home/tduong/NetApp/oncommand_system_manager/SystemManager.jar &") 
	, ((myModMask .|. shiftMask,	xK_n),		spawn "nixnote")
	, ((myModMask,                  xK_q),          spawn "killall xpad shutter stalonetray xmobar nm-applet gnome-setting-daemon xautolock" >> restart "xmonad" True)
	, ((myModMask,                  xK_r),          spawn "remmina")
	, ((myModMask,                  xK_x),          spawn "nxclient")
        , ((myModMask .|. shiftMask,  	xK_r),          spawn "/usr/local/bin/vncviewer")
--	, ((myModMask .|. shiftMask,	xK_r),          spawn "/usr/bin/vnc-E4_6_3-x64_linux_viewer")
	, ((myModMask,                  xK_comma), 	sendMessage (IncMasterN 1))
	, ((myModMask,                  xK_period), 	sendMessage (IncMasterN (-1)))
 	, ((controlMask,		xK_Print),      spawn "/home/tduong/.xmonad/bin/select-sreenshot")
--	, ((0,                          xK_Print),      spawn "/home/tduong/.xmonad/bin/screenshot") 
      	, ((myModMask,                  xK_Print),      spawn "shutter -s")
        , ((0, 				xK_Print), 	spawn "shutter -f")
 	--Volume control
	-- Mute toggle
	, ((0 , xF86XK_AudioMute),                      spawn "amixer -q set Master toggle")
--	, ((0 , xF86XK_MicMute),			spawn "amixer -q set Master toggle")
	, ((0 , xF86XK_AudioLowerVolume),               spawn "amixer -q set Master 10%-")
	, ((0 , xF86XK_AudioRaiseVolume),               spawn "amixer -q set Master 10%+")
--	, ((0 , xF86XK_AudioLowerVolume),               spawn "amixer -c0 set Master on && amixer -c0 set Headphone on && amixer -c0 set Master 5%-")
--	, ((0 , xF86XK_AudioRaiseVolume),               spawn "amixer -c0 set Master on && amixer -c0 set Headphone on && amixer -c0 set Master 5%+")
--	, ((0 , xF86XK_AudioMute), 			spawn "amixer set Master toggle && amixer set Headphone toggle")
	-----------------------------
        , ((myModMask .|. controlMask,   xK_l),         spawn "xscreensaver-command -lock")
        , ((myModMask,                   xK_v),         spawn "virtualbox &")
        , ((myModMask .|.controlMask,    xK_v),         spawn "wine '/home/tduong/.wine/drive_c/Program Files (x86)/Visual CertExam Suite/manager.exe'")
        ]
--- Added more application "ipcalc, eog, scribus, inkscape, evince, foxitpdf, okular
-- The Main Config --
main = do
    spawn "ibus-daemon"
    spawn "xautolock"
    spawn "xpad"
    spawn "gnome-settings-daemon"
    spawn "xcompmgr -c"
    spawn "nm-applet --sm-disable"
    spawn "stalonetray"
    spawn "feh --bg-scale ~/.wallpapers/desktop.jpg"
    spawn "/opt/dropbox/dropbox &"
    xmproc <- spawnPipe "/usr/bin/xmobar ~/.xmonad/xmobarrc"
    xmonad $ defaultConfig { 
        manageHook = manageDocks <+> manageHook defaultConfig <+> composeAll myManagementHooks 
        , layoutHook = avoidStruts  $ smartBorders $ myLayout 
	, handleEventHook = ewmhDesktopsEventHook
        , logHook = myLogHook <+> ewmhDesktopsLogHook <+> dynamicLogWithPP xmobarPP
                { ppOutput = hPutStrLn xmproc
                , ppTitle = xmobarColor "blue" "" . shorten 50
                }
        , terminal      = myTerminal
        , modMask       = myModMask
        , borderWidth   = myBorderWidth
        , workspaces    = myWorkspaces
        , keys          = myKeys <+> keys defaultConfig
        , focusFollowsMouse = False
}
